### PR TITLE
Rename overloaded attribute method to remove ambiguity from extension method

### LINKED ISSUE #19014: Remove ambiguity in overloaded AttributeContainer#attribute methods
The new overload for `AttributeContainer#attribute` caused ambiguity when the `ObjectFactoryExtensions` file's `named` method is used.

Though you can resolve this by adding a generic type argument to any calls to the method, this has been show to be too large a breaking change for the next minor version.

- Rename the new method to `attributeProvider`.
- Remove the section in the upgrade guide regarding this issue.

### PR DESCRIPTION
**Pull request by tresat**
Fixes #19014 

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by melix** [comment]
On a related topic this line should probably be changed: https://github.com/gradle/gradle/blob/94c767928ea4ce0f233f473fed968e45de2359e1/subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java#L36

The attribute containers proved to be extremely memory hungry and an effort was made to reduce allocations as much as possible. It would therefore be good to evaluate the impact of this change which creates an empty map even if no lazy attribute is set.

**Comment by bamboo** [comment]
> The attribute containers proved to be extremely memory hungry and an effort was made to reduce allocations as much as possible. It would therefore be good to evaluate the impact of this change which creates an empty map even if no lazy attribute is set.

CHAMP maps could help a lot here.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
What do you think about the latest commit - using an `ImmutableMap`to minimize storage - especially for the common empty case?

**Comment by melix** [comment]
I would probably have done something simpler: use a `null` field and a `LinkedHashMap` as soon as something is added: otherwise you're making addition quite expensive and creating lots of intermediate objects (creates a new map, duplicates to immutable map). But without measuring with a profiler it's hard to tell, I just remember that when I worked on that, the dumps for attribute containers were huge, so we had to find "smart" ideas...

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 1]
```diff
@@ -209,4 +210,22 @@ public int hashCode() {
...
+        @SuppressWarnings("unchecked")
+        final Provider<? extends T> result = (Provider<? extends T>) lazyAttributes.get(key);
+
+        final Map<Attribute<?>, Provider<?>> temp = Maps.newHashMap(lazyAttributes);
+        //noinspection SuspiciousMethodCalls
+        temp.remove(lazyAttributes);
+        lazyAttributes = ImmutableMap.copyOf(temp);
+
+        return result;
+    }
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:230]: Do we really get anything out of using an immutable map here?  We end up just creating a bunch of new maps vs updating a single instance?

Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:230]: Just trying to use a different `Map` implementation to minimize storage, and was influenced by the pattern used for storing state.  I'll switch to a pre-sized `LinkedHashMap`.

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java (hunk 1 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java]: > Do we really get anything out of using an immutable map here?

We always get good stuff from using immutable data structures. In this particular case:
* no `null` checks anywhere, no branching
* maps can be immediately shared without copying

> We end up just creating a bunch of new maps vs updating a single instance?

Lots of assumptions here :)

Consider, for instance, that CHAMP data structures have lower memory overhead than JDK collections.

These Guava ImmutableMap with copying behavior provide some nice properties but are, in general, much less sophisticated implementations of these concepts.

While I understand the reluctance to applying CHAMP here (lack of familiarity) I still think it's the superior general solution.

**Comment by big-guy** [comment]
@bot-gradle test QF

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java (hunk 1 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java]: > ...  I think we could do something simple here with an empty map sentinel value vs using null. ...

Sounds good to me 👍

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@bot-gradle cancel

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**